packages/shared/src/__tests__/retry.test.ts

```diff
@@ -0,0 +1,209 @@
+    await jest.advanceTimersByTimeAsync(300);
+    await jest.advanceTimersByTimeAsync(300);
+    await jest.advanceTimersByTimeAsync(300);
+    expect(attempts).toBe(1 + 4);
+    jest.useRealTimers();
+  });
+
+  test('respects firstDelay option', async () => {
+    let attempts = 0;
+    retry(
+      () => {
+        attempts++;
+        throw new Error('failed');
+      },
+      { firstDelay: 200, jitter: false, shouldRetry: (_, count) => count <= 2 },
+    ).catch(() => {});
+
+    expect(attempts).toBe(1);
+    await jest.advanceTimersByTimeAsync(200);
+    expect(attempts).toBe(2);
+    await jest.advanceTimersByTimeAsync(400);
+    expect(attempts).toBe(3);
+  });
+
+  test('respects retryImmediatelyOnce option', async () => {
+    let attempts = 0;
+    retry(
+      () => {
+        attempts++;
+        throw new Error('failed');
+      },
+      {
+        firstDelay: 1000,
+        retryImmediatelyOnce: true,
+        retryImmediatelyDelay: 50,
+        jitter: false,
+        shouldRetry: (_, count) => count <= 2,
+      },
+    ).catch(() => {});
+
+    expect(attempts).toBe(1);
+    await jest.advanceTimersByTimeAsync(50);
+    expect(attempts).toBe(2);
+    await jest.advanceTimersByTimeAsync(1000);
+    expect(attempts).toBe(3);
+  });
+
+  test('disables immediate retry when retryImmediatelyOnce is false', async () => {
+    let attempts = 0;
+    retry(
+      () => {
+        attempts++;
+        throw new Error('failed');
+      },
+      {
+        firstDelay: 200,
+        retryImmediatelyOnce: false,
+        jitter: false,
+        shouldRetry: (_, count) => count <= 2,
+      },
+    ).catch(() => {});
+
+    expect(attempts).toBe(1);
+    await jest.advanceTimersByTimeAsync(200);
+    expect(attempts).toBe(2);
+    await jest.advanceTimersByTimeAsync(400);
+    expect(attempts).toBe(3);
+  });
+
+  test('respects shouldRetry custom logic', async () => {
+    let attempts = 0;
+    const error = new Error('special error');
+
+    await retry(
+      () => {
+        attempts++;
+        throw error;
+      },
+      {
+        firstDelay: 100,
+        jitter: false,
+        shouldRetry: e => (e as Error).message !== 'special error',
+      },
+    ).catch(e => {
+      expect(e).toBe(error);
+    });
+
+    expect(attempts).toBe(1);
+  });
+
+  test('respects factor for exponential backoff', async () => {
+    let attempts = 0;
+    retry(
+      () => {
+        attempts++;
+        throw new Error('failed');
+      },
+      {
+        firstDelay: 100,
+        factor: 4,
+        jitter: false,
+        shouldRetry: (_, count) => count <= 3,
+      },
+    ).catch(() => {});
+
+    expect(attempts).toBe(1);
+    await jest.advanceTimersByTimeAsync(100);
+    expect(attempts).toBe(2);
+    await jest.advanceTimersByTimeAsync(400);
+    expect(attempts).toBe(3);
+    await jest.advanceTimersByTimeAsync(1600);
+    expect(attempts).toBe(4);
+  });
+
+  test('applies jitter when enabled', async () => {
```

Comment:
```suggestion
  test('applies jitter when enabled by default', async () => {
```

Suggestion: Indicate that it's enabled by default on the description, since the `// jitter: true` below was a bit confusing to me if I haven't read the PR description / implementation

Reply:
Great suggestion Laura, will change the test accordingly!